File: test/elixir/hologram/ex_js_consistency/erlang/os_test.exs
defmodule Hologram.ExJsConsistency.Erlang.OsTest do
  @moduledoc """
  IMPORTANT!
  Each Elixir consistency test has a related JavaScript test in test/javascript/erlang/os_test.mjs
  Always update both together.
  """

  use Hologram.Test.BasicCase, async: true

  @moduletag :consistency

  describe "system_time/0" do
    test "returns current system time in native time unit (nanoseconds)" do
      before_ns = :os.system_time(:nanosecond)
      result = :os.system_time()
      after_ns = :os.system_time(:nanosecond)

      assert is_integer(result)
      assert result >= before_ns
      assert result <= after_ns
    end
  end

  describe "system_time/1" do
    test "with valid atom unit" do
      assert is_integer(:os.system_time(:second))
    end

    test "with valid integer unit" do
      assert is_integer(:os.system_time(1000))
    end

    test "applies time unit conversion" do
      micro = :os.system_time(:microsecond)
      nano = :os.system_time(:nanosecond)

      # Allow small timing drift between calls
      assert nano >= micro * 999
      assert nano <= micro * 1001 + 1000
    end

    test "raises ArgumentError when argument is not atom or integer" do
      assert_error ArgumentError,
                   build_argument_error_msg(1, "invalid time unit"),
                   {:os, :system_time, [1.0]}
    end

    test "raises ArgumentError when atom argument is not a valid time unit" do
      assert_error ArgumentError,
                   build_argument_error_msg(1, "invalid time unit"),
                   {:os, :system_time, [:invalid]}
    end

    test "raises ArgumentError when integer argument is 0" do
      assert_error ArgumentError,
                   build_argument_error_msg(1, "invalid time unit"),
                   {:os, :system_time, [0]}
    end

    test "raises ArgumentError when integer argument is negative" do
      assert_error ArgumentError,
                   build_argument_error_msg(1, "invalid time unit"),
                   {:os, :system_time, [-1]}
    end
  end

  describe "type/0" do
    test "returns OS family and OS name" do
      assert {family, name} = :os.type()

      assert family in [:win32, :unix]
      assert is_atom(name)
    end
  end
end

File: test/javascript/erlang/os_test.mjs
"use strict";

import {
  assert,
  assertBoxedError,
  defineGlobalErlangAndElixirModules,
} from "../support/helpers.mjs";

import Erlang_Os from "../../../assets/js/erlang/os.mjs";
import Interpreter from "../../../assets/js/interpreter.mjs";
import Type from "../../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

// IMPORTANT!
// Each JavaScript test has a related Elixir consistency test in test/elixir/hologram/ex_js_consistency/erlang/os_test.exs
// Always update both together.

describe("Erlang_Os", () => {
  describe("system_time/0", () => {
    const systemTime = Erlang_Os["system_time/0"];

    it("returns current system time in native time unit (nanoseconds)", () => {
      const beforeMs = Date.now();
      const result = systemTime();
      const afterMs = Date.now();

      assert.isTrue(Type.isInteger(result));
      assert.isAtLeast(result.value, BigInt(beforeMs) * 1_000_000n);
      assert.isAtMost(result.value, BigInt(afterMs + 1) * 1_000_000n);
    });
  });

  describe("system_time/1", () => {
    const systemTime = Erlang_Os["system_time/1"];

    it("with valid atom unit", () => {
      const result = systemTime(Type.atom("second"));

      assert.isTrue(Type.isInteger(result));
    });

    it("with valid integer unit", () => {
      const result = systemTime(Type.integer(1000));

      assert.isTrue(Type.isInteger(result));
    });

    it("applies time unit conversion", () => {
      const micro = systemTime(Type.atom("microsecond")).value;
      const nano = systemTime(Type.atom("nanosecond")).value;

      // Allow small timing drift between calls
      assert.isTrue(nano >= micro * 999n);
      assert.isTrue(nano <= micro * 1001n + 1000n);
    });

    it("raises ArgumentError when argument is not atom or integer", () => {
      assertBoxedError(
        () => systemTime(Type.float(1.0)),
        "ArgumentError",
        Interpreter.buildArgumentErrorMsg(1, "invalid time unit"),
      );
    });

    it("raises ArgumentError when atom argument is not a valid time unit", () => {
      assertBoxedError(
        () => systemTime(Type.atom("invalid")),
        "ArgumentError",
        Interpreter.buildArgumentErrorMsg(1, "invalid time unit"),
      );
    });

    it("raises ArgumentError when integer argument is 0", () => {
      assertBoxedError(
        () => systemTime(Type.integer(0)),
        "ArgumentError",
        Interpreter.buildArgumentErrorMsg(1, "invalid time unit"),
      );
    });

    it("raises ArgumentError when integer argument is negative", () => {
      assertBoxedError(
        () => systemTime(Type.integer(-1)),
        "ArgumentError",
        Interpreter.buildArgumentErrorMsg(1, "invalid time unit"),
      );
    });
  });

  describe("type/0", () => {
    const type = Erlang_Os["type/0"];

    it("returns OS family and OS name", () => {
      const result = type();

      assert.deepStrictEqual(
        result,
        Type.tuple([Type.atom("unix"), Type.atom("web")]),
      );
    });
  });
});
